build-plugins/build-support/src/main/kotlin/aws/sdk/kotlin/gradle/plugins/artifactmetrics/AnalyzeArtifactSizeMetricsTask.kt

```diff
@@ -0,0 +1,166 @@
+        writeLatestReleaseMetrics(latestReleaseMetricsFile)
+
+        val latestReleaseMetrics = latestReleaseMetricsFile.toMap()
+        val currentMetrics = metricsFile.get().asFile.toMap()
+        val analysis = analyzeArtifactSizeMetrics(latestReleaseMetrics, currentMetrics)
+
+        hasSignificantChangeFile.get().asFile.writeText(analysis.significantChange.toString())
+        val diffTable = createDiffTable(analysis)
+        val output = if (analysis.delta) diffTable else noDiffMessage
+
+        this.analysisFile.get().asFile.writeText(output)
+    }
+
+    private fun writeLatestReleaseMetrics(file: File) = runBlocking {
+        S3Client.fromEnvironment().use { s3 ->
+            s3.getObject(
+                GetObjectRequest {
+                    bucket = "" // TODO: Point to artifact size metrics bucket
+                    key = "" // TODO: Point to artifact size metrics for latest release
+                },
+            ) { latestReleaseMetrics ->
+                file.writeText(
+                    latestReleaseMetrics.body?.decodeToString() ?: throw GradleException("Metrics from latest release are empty"),
+                )
+            }
+        }
+    }
+
+    private fun analyzeArtifactSizeMetrics(
+        releaseMetrics: Map<String, Long>,
+        currentMetrics: Map<String, Long>,
+    ): ArtifactSizeMetricsAnalysis {
+        val pluginConfig = this.project.rootProject.extensions.getByType(ArtifactSizeMetricsPluginConfig::class.java)
+
+        val artifactNames = releaseMetrics.keys + currentMetrics.keys
+        val artifactSizeMetrics = artifactNames.associateWith { artifact ->
+            val current = currentMetrics[artifact] ?: 0
+            val release = releaseMetrics[artifact] ?: 0
+
+            val delta = current - release
+            val percentage = if (current == 0L || release == 0L) Double.NaN else delta.toDouble() / release.toDouble() * 100
+
+            ArtifactSizeMetric(
+                current,
+                release,
+                delta,
+                percentage,
+            )
+        }
+
+        val changeHappened = artifactSizeMetrics.values.any { it.delta != 0L }
+        val significantChange = artifactSizeMetrics.values.any {
+            (abs(it.percentage) > pluginConfig.significantChangeThresholdPercentage && it.delta > 0L) || // Increase in size above threshold
+                (it.latestReleaseSize == 0L) // New artifact
+        }
+
+        return ArtifactSizeMetricsAnalysis(artifactSizeMetrics, significantChange, changeHappened)
+    }
+
+    private data class ArtifactSizeMetricsAnalysis(
+        val metrics: Map<String, ArtifactSizeMetric>,
+        val significantChange: Boolean,
+        val delta: Boolean,
+    )
+
+    private fun createDiffTable(analysis: ArtifactSizeMetricsAnalysis): String = buildString {
+        appendLine("Affected Artifacts\n=")
```

Comment:
question/style: Should there be another `\n` after the `=` to match the format of `noDiffMessage`?

Reply:
The `appendLine(...)` adds a `\n` to the end of the string so it matches the format